### ISSUE TITLE
`knife cookbook upload` not respecting cookbook chefignore file

### ISSUE THREAD

**Issue by shoekstra** [chef/chef#8109]
Ohai!

It seems in recent version of chef that `knife` (and `berks`) no longer respect the `chefignore` file in a cookbook directory when uploading cookbooks to a chef server.

I use chef-guard to enforce a cookbook upload contains the same content as it's supermarket artifact when freezing.  It does this by fetching the artifact from the supermarket and compare what is uploaded by `knife` to the artifact for differences.

When trying to freeze with recent chef versions I now receive errors:

```
$ knife cookbook upload certificate_services --freeze
Uploading certificate_services [0.1.0]
ERROR: Precondition Failed
Response:
=== Cookbook Compare errors found ===
Your upload contains more files than the source cookbook:
 - Gemfile
 - LICENSE
 - chefignore
 - tags

Source: https://supermarket.chef.io/api/v1/cookbooks/certificate_services/versions/0.1.0/download
=====================================

$ rm Gemfile LICENSE chefignore tags
$ knife cookbook upload certificate_services --freeze
Uploading certificate_services [0.1.0]
ERROR: Precondition Failed
Response:
=== Cookbook Compare errors found ===
Your upload contains more files than the source cookbook:
 - .gitignore
 - .kitchen.yml
 - .rspec
 - Berksfile

Source: https://supermarket.chef.io/api/v1/cookbooks/certificate_services/versions/0.1.0/download
=====================================

$ rm .gitignore .kitchen.yml .rspec Berksfile
$ knife cookbook upload certificate_services --freeze
Uploading certificate_services [0.1.0]
ERROR: Precondition Failed
Response:
=== Cookbook Compare errors found ===
Your upload contains more files than the source cookbook:
 - Berksfile.lock

Source: https://supermarket.chef.io/api/v1/cookbooks/certificate_services/versions/0.1.0/download
=====================================

$ rm Berksfile.lock
$ knife cookbook upload certificate_services --freeze
Uploading certificate_services [0.1.0]
Uploaded 1 cookbook.
```

All mentioned files are matched in my `chefignore` which leads me to believe something has changed in the cookbook upload process.

**Comment by vijaymmali1990** [comment]
Could you please let me know the Chef version you are using? I tried on the latest version of Chef and it's working fine for me. Also please provide me knife configrations and chefignore file content.
Output:

Knife cookbook upload:

```
ABC@PUN-DSK-105:~/Chef/Employee/chef-repo$ knife cookbook upload certificate_services --freeze
Uploading certificate_services [0.1.0]
Uploaded 1 cookbook.
```

Chef -v 

```
Chef Development Kit Version: 3.7.23
chef-client version: 14.10.9
berks version: 7.0.7
kitchen version: 1.24.0
inspec version: 3.4.1
```

**Comment by shoekstra** [comment]
Hi,

As requested:

```
Chef Development Kit Version: 3.6.57
chef-client version: 14.8.12
delivery version: master (5fb4908da53579c9dcf4894d4acf94d2d9ee6475)
berks version: 7.0.7
kitchen version: 1.24.0
inspec version: 3.2.6
```

Interestingly when I download the artifact from the supermarket and from the Chef server they both look the same, neither contains the files that are mentioned above so not sure why Chef-Guard is complaining that `knife` is uploading more files when comparing to the supermarket artifact.

I was under the assumption that `chefignore` was being ignored when the upload was happening using `knife`, but maybe all files are uploaded the chef server disregards what's in the `chefignore`?

I'll do some more tests and get back to you - thanks for the response!

**Comment by vijaymmali1990** [comment]
Hi I tried working on chef 3.6.57 and it works fine for me here too.

**Comment by Plork** [comment]
I don't think you will see anything if you dont have chef-guard installed on your chef server.

http://xanzy.io/projects/chef-guard/introduction/overview.html

**Comment by shoekstra** [comment]
Hi @vijaymmali1990,

I've been digging into this a bit more and noticed there is a regression.

Testing with the latest release of the HAProxy cookbook:

```
$ knife cookbook upload haproxy --freeze -VVV >! /tmp/haproxy_upload.txt
Uploading haproxy        [6.4.0]
ERROR: Precondition Failed
Response:
=== Cookbook Compare errors found ===
Your upload contains more files than the source cookbook:
 - .envrc
 - .mdlrc
 - .yamllint
 - CODE_OF_CONDUCT.md
 - Dangerfile
 - LICENSE
 - chefignore

Source: https://supermarket.chef.io/api/v1/cookbooks/haproxy/versions/6.4.0/download
=====================================
```

The chefignore file is indeed being ignored, thus knife tries to upload more files than are in the supermarket which triggers the above.

```
$ grep chefignore /tmp/haproxy_upload.txt
TRACE: No chefignore file found at /Users/shoekstra/chefignore no files will be ignored
TRACE: No chefignore file found at /Users/shoekstra/cookbooks/cookbooks/chefignore no files will be ignored
```

The two locations checked should actually be `/Users/shoekstra/cookbooks/haproxy/chefignore` and `/Users/shoekstra/cookbooks/chefignore`

Thanks,
Stephen

**Comment by knightorc** [comment]
@shoekstra @Plork 

Hitting the same issue.  We have a mono-repo.

We have two cookbook folders

`com` == community cookbooks
`ours` == ours or our wrapper cookbooks

IE

```
/path/to/repo/com/nginx
/path/to/repo/ours/wrapper_nginx
```

`-VVV` output has like

```
TRACE: No chefignore file found at /path/to/repo/com/path/to/repo/ours/chefignore
```
So it never traverses into each cookbook folder.

**Comment by knightorc** [comment]
Tested with `14.8.12`, 14.11.21` and latest 14.12.9`

**Comment by shoekstra** [comment]
Hey @knightorc, 

Just tested with `15.0.300` and issue still exists.  I have a case open with Chef support about this so hope to get it sorted soonest 👍

**Comment by vsingh-msys** [comment]
Fixed in #8925
